#455217 [C#テキスト]1章演習問題

kurosu-being opened the pull request.

kurosu-being committed:
章末問題第1章1.1-1～1.2-1までを実装しました

kurosu-being committed:
演習問題第1章1.2-2～1.3-4までを実装しました。

kurosu-being committed:
#455217 [C#テキスト]1章演習問題

メンバー変数などの名付けを一部変更しました。

k-miyahara-being commented on 章末問題第1章/章末問題第1章/Program.cs:
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Henkou;
```
不要なusingは消しておきましょう。
（文字色がグレーになっているものが、使われていないことを表しています。）

ショートカットキー(Ctrl + K, Ctrl + D)で、開いているファイル全体のコード整形を自動でやってくれます。
使われていないusingは自動でまとめて削除してくれたり、ずれているインデントがあったら修正してくれたりします。
とにかく見た目をきれいにしてくれる機能と思ってくれたらよいです。
コードを書ききったタイミングで毎回やる癖がついていると、きれいなコードを保つことができます！
見た目がきれいなコードを書くことも意識してみてください。

ちなみに、選択範囲だけコード整形したいときは、(Ctrl + K, Ctrl + F)を使用します。


他の同様の箇所も修正してください。

k-miyahara-being commented on 章末問題第1章/章末問題第1章/Program.cs:
```diff
@@ -0,0 +1,17 @@
+        //Practice1.1-1 どら焼きオブジェクトを生成するコードを書いてください。この時の商品番号は98、商品価格は"210円"としてください。
+        //Practice1.1-2　どら焼きオブジェクトの消費税額を求め、コンソールに出力するコードを書いてください
+        static void Main(string[] args) {
+            Product dorayaki = new Product(98, "どら焼き", 210);
```
dorayaki はローカル変数なので、コーディング規約にあてはめて書き換えてみてください。
他の同様の箇所も修正してください。

k-miyahara-being commented on 章末問題第1章/章末問題第1章/Product.cs:
```diff
@@ -0,0 +1,33 @@
+using System.Threading.Tasks;
+
+namespace Henkou {
+    public class Product {
+        //商品コード
+        public int Code { get; private set; }
+        //商品名
+        public string Name { get; private set; }
+        //商品価格（税抜き）
+        public int Price{ get; set;}
```
クラス、プロパティにそれぞれヘッダーコメントを付けてみてください。
1つ上の行で`///`と入力するとヘッダーコメントが現れるので、そのクラス、プロパティが何を表すものなのかを日本語で補足してください。
プロパティの説明は、今書いてあるような感じでOKです！
ヘッダーコメントを付けると↓みたいになります。
```
/// <summary>
/// 商品コード
/// </summary>
public int Code { get; private set; }
```

ヘッダーコメントがあると、第三者が見たときに、どのクラス、プロパティを使えば良いか適切に判断できるようになります。
またVisual Studioの機能として、クラス名、プロパティ名にマウスホバーしたとき、ヘッダーコメントに書いた説明が見れるようになるので、わざわざ定義に移動しなくて済み便利になります。

ほかに、publicなメソッドにもヘッダーコメントをつけるようにしてください。


参考：[Wiki>コーディング規約>コメント](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#%E3%82%B3%E3%83%A1%E3%83%B3%E3%83%88)

k-miyahara-being commented on 章末問題第1章/章末問題第1章/Product.cs:
```diff
@@ -0,0 +1,33 @@
+            this.Price = price;
+        }
+
+        //消費税を求める（消費税率は8%）
+        public int GetTax(){
+            return (int)(Price * 0.1);
+        }
```
コメントと処理の中身が一致していませんね
実際の処理では、消費税率が10%になっています。

k-miyahara-being commented on 章末問題第1章/章末問題第1章/Program.cs:
```diff
@@ -0,0 +1,17 @@
+        //Practice1.1-2　どら焼きオブジェクトの消費税額を求め、コンソールに出力するコードを書いてください
+        static void Main(string[] args) {
+            Product dorayaki = new Product(98, "どら焼き", 210);
+            Console.WriteLine("消費税額は" + dorayaki.GetTax());
```
単位もつけて出力してみてください！

k-miyahara-being reviewed (changes requested):
問1-1、1-2のレビューをしました。
修正してみてください。

k-miyahara-being commented on 章末問題第1章/Practice1.2-1/Program.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Practice1._2_1 {
+    class Program {
+        //Practice1.2-1　MyClassとMyStructの2つの方を引数にとるメソッドPrintObjectsを定義してください。PrintObjectsメソッドでは、2つのオブジェクトの内容（プロパティの値)をコンソールに表示するようにしてください。なお、PrintObjectsメソッドはProgramクラスのメソッドとして定義してください。
```
方→型ですね😁

k-miyahara-being commented on 章末問題第1章/Practice1.2-1/Program.cs:
```diff
@@ -0,0 +1,24 @@
+            MyStruct myStruct = new MyStruct { X = 1, Y = 2 };
+            PrintObjects(myClass,myStruct);
+            Console.ReadLine();
+        }
+        public static void PrintObjects(MyClass vMC, MyStruct vMS) {
+            Console.WriteLine("X =" + vMC.X +"Y =" + vMC.Y);
+            Console.WriteLine("X =" + vMS.X + "Y =" + vMS.Y);
+
+        }
+       
+    }
```
細かいですが、
メソッド間(16, 17行目の間)は空行を入れると見通しがよくなります。
また、20行目や22行目の空行は不要なので削除しておいてください。

k-miyahara-being reviewed (changes requested):
問2-1のレビューをしました。
大体OKです！

k-miyahara-being commented on 章末問題第1章/Practice1.1-3/Program.cs:
このファイルは必要ですか？
必要なければ削除してください。

k-miyahara-being commented on 章末問題第1章/Practice1.3-3, 1.3-4/Student.cs:
```diff
@@ -0,0 +1,13 @@
+using System.Threading.Tasks;
+
+namespace Practice1._3_1 {
+     class Student : Person {
+        public int Grade { get; set; }
+        public int ClassNumber{ get; set;}
+
+    }
```
Studentクラスにコンストラクタを実装して、そこでプロパティを設定するようにしてみてください。

k-miyahara-being commented on 章末問題第1章/Practice1.3-1/Program.cs:
```diff
@@ -0,0 +1,19 @@
+            Student student = new Student{
+                Name = "黒子", Birthday = new DateTime(2000, 2, 2),Grade = 3, ClassNumber = 10,
+            };
+            Console.WriteLine("{0}は、{1}年{2}組に所属しています",student.Name, student.Grade, student.ClassNumber);
```
`"{0}は、{1}年{2}組に所属しています",student.Name, student.Grade, student.ClassNumber`
この部分を、文字列補間式で書き換えてみてください。
参考：https://csharp.keicode.com/basic/string-interpolation.php

k-miyahara-being reviewed (changes requested):
問3-1のレビューをしました。

kurosu-being committed:
455217 [C#テキスト]1章演習問題

全てのcsにおいて、不要なusingを消去し、コードを整形しました。
全てのPrgram.csをそれぞれのファイル名に対応した名前に変更しました。（例：Program.cs → Program1.1-1,1.1-2.1.1-3.cs）また、その際不要な章末問題1.1-3ファイルを削除し、第1章1.1の問題をすべて同じファイルに統合しました。
Practice1.1,1.1-2,1.1-3において、クラス、プロパティにそれぞれにヘッダーコメントを付けました。
Practice1.1,1.1-2,1.1-3において、ローカル変数の命名をdorayaki→wdorayakiに変更しました。また、以降問題についてもmyClass,mystudentを→wmyClass,wstudentに変更しました。

kurosu-being committed:
455217 [C#テキスト]1章演習問題

全てのcsにおいて、不要なusingを消去し、コードを整形しました。
全てのPrgram.csをそれぞれのファイル名に対応した名前に変更しました。（例：Program.cs → Program1.1-1,1.1-2.1.1-3.cs）また、その際不要な章末問題1.1-3ファイルを削除し、第1章1.1の問題をすべて同じファイルに統合しました。
Practice1.1,1.1-2,1.1-3において、クラス、プロパティにそれぞれにヘッダーコメントを付けました。
Practice1.1,1.1-2,1.1-3において、Console.WriteLineに"円です"が表示されるように追加し、結果の単位が表示されるように修正しました。
Practice1.1,1.1-2,1.1-3において、ローカル変数の命名をdorayaki→wdorayakiに変更しました。また、以降問題についてもmyClass,mystudentを→wmyClass,wstudentに変更しました。

m-hotta-being commented on 章末問題第1章/Practice1.2-1/Practice1.2-1,1.2-2.csproj:
> /Practice1.2-1/Priactice1.2-1,1.2-2.csproj

一般的に、ファイル名に `.` や `,` などを使用するのはオススメしません。
無用なトラブルを避けるためです。
参考： <https://ascii.jp/elem/000/004/030/4030753/>

私なら次のようにするかなぁ。
> /Practice1-2-1/Practice1-2-1_1-2-2.csproj

kurosu-being replied:
ご指摘ありがとうございました！
参考URL、とても勉強になりました。分かりやすいファイル名の名付けを心がけたいと思います。

m-hotta-being commented on 章末問題第1章/章末問題第1章/Program1.1.cs:
```diff
@@ -0,0 +1,14 @@
+﻿using Henkou;
+using System;
```
using の並び順が逆ですね。
参考：[コーディング規約 - using](https://dev-redmine.being.group/projects/sirius-client/wiki/%E3%82%B3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0%E8%A6%8F%E7%B4%84#using)

kurosu-being replied:
ご指摘ありがとうございました！
コーディング規約通り.NET標のSystemを一番上に変更しました。

m-hotta-being commented on 章末問題第1章/章末問題第1章/Program1.1.cs:
```diff
@@ -0,0 +1,14 @@
+    class Program {
+        //Practice1.1-1 どら焼きオブジェクトを生成するコードを書いてください。この時の商品番号は98、商品価格は"210円"としてください。
+        //Practice1.1-2　どら焼きオブジェクトの消費税額を求め、コンソールに出力するコードを書いてください
+        //Productクラスが属する名前空間を別の名前空間に変更し、Mainメソッドから呼び出すようにしてください。ただし、MainメソッドのあるProgramクラスの名前空間はそのままとしてください。
```
【修正不要。コメントです。】
コメントには次のように `/* */` で囲うスタイルもあります。
あまり使わないと思いますが、覚えておいて損はないと思います。
```
/*
コメント
*/
```

kurosu-being replied:
勉強になります！
上と下で挟むコメントで統一出来るという点で分かりやすいですね。
ありがとうございます！

m-hotta-being reviewed (changes requested):
ちょろっと見てみました。
ファイル名については早いうちに知っておいたほうがよいと思うので、参考につけたリンク先の記事も読んでみてください。

kurosu-being committed:
#455217 [C#テキスト]1章演習問題

全てのProgram.csにおいて、問題の表記方法を変更しました。（例　Program1.2-1,1.2-2→Program1-2-1,1-2-2）
Program1-1-1,1-1-2,1-1-3において、usingの並びを変更しました。
Program1-2-1,1-2-2において、問題文の修正を行いました（方→型）
Program1-2-1,1-2-2において、メソッドの間に空行を追加しました。また、空行の削除を行いました。
問３においてStudent.csの
"{0}は、{1}年{2}組に所属しています",student.Name, student.Grade, student.ClassNumber
を文字列補間式で書き換えました。

kurosu-being committed:
Merge remote-tracking branch 'origin/feature/chapter1' into feature/chapter1

# Conflicts:
#	章末問題第1章/Practice1.2-1/Practice1-2-1,1-2-2.csproj
#	章末問題第1章/Practice1.2-1/Practice1.2-1,1.2-2.csproj
#	章末問題第1章/Practice1.2-1/Practice1.2-1.csproj
#	章末問題第1章/Practice1.2-1/Program.cs
#	章末問題第1章/Practice1.2-1/Program1-2-1,1-2-2.cs
#	章末問題第1章/Practice1.2-1/Program1.2-1,1.2-2.cs
#	章末問題第1章/Practice1.2-3/Practice1-2-3.csproj
#	章末問題第1章/Practice1.2-3/Program.cs
#	章末問題第1章/Practice1.2-3/Program1-2-3.cs
#	章末問題第1章/Practice1.2-3/Program1.2-3.cs
#	章末問題第1章/Practice1.3-1/Practice1-3-1, 1-3-2.csproj
#	章末問題第1章/Practice1.3-1/Program.cs
#	章末問題第1章/Practice1.3-1/Program1-3-1,1-3-2.cs
#	章末問題第1章/Practice1.3-1/Program1.3-1,1.3-2.cs
#	章末問題第1章/Practice1.3-1/Student.cs
#	章末問題第1章/Practice1.3-3, 1.3-4/Practice1-3-4.csproj
#	章末問題第1章/Practice1.3-3, 1.3-4/Program1.3-4.cs
#	章末問題第1章/Practice1.3-3, 1.3-4/Student.cs
#	章末問題第1章/章末問題第1章.sln
#	章末問題第1章/章末問題第1章/Practice1-1-1, 1-1-2,1-1-3.csproj
#	章末問題第1章/章末問題第1章/Practice1.1-1, 1.1-2,.csproj
#	章末問題第1章/章末問題第1章/Practice1.1-1, 1.1-2,1-3.csproj
#	章末問題第1章/章末問題第1章/Program1.1.cs

s-morita-being commented on 章末問題第1章/章末問題第1章/Product.cs:
```diff
@@ -0,0 +1,34 @@
+        /// /// </summary>
+        /// <returns>int</returns>
+        public int GetTax() {
+            return (int)(Price * 0.1);
```
演習問題を見てませんが、実際のお金の計算では端数をどう扱うかが決まっているはずなので、それに合わせた丸めをする必要があります。
`Price` が 15 円の場合の消費税は? 14 円の場合は? 16 円の場合は?
みたいな。

kurosu-being replied:
ご指摘ありがとうございます！
演習問題では特に指定はありませんでしたが確かに、小数点以下の処理が必要だと思います。
Math.roundメソッドで小数点以下を四捨五入して返すように変更いたしました。

s-morita-being commented on 章末問題第1章/章末問題第1章/Product.cs:
```diff
@@ -0,0 +1,34 @@
+﻿namespace Henkou {
+    public class Product {
+        /// <summary>
+        /// 商品コード
+        /// </summary>
+        public int Code { get; private set; }
+        /// <summary>
+        /// <商品名>
+        /// </summary>
+        public string Name { get; private set; }
+        /// <summary>
+        /// <商品価格（税抜き）>
+        /// </summary>
+        public int Price { get; set; }
```
コンストラクタで初期化したら変更されないプロパティの場合は `get;` だけでよいです。

各フィールド/プロパティの定義を考えるときに、
一度決めたら変更されないものなのか、後から変更できるのか、第三者が変更できるのか、
などを意識するとよいです。

後から変更できるものをなるべく少なくして、変更する方法も制限できると、
壊れにくいプログラムにしやすいと思います。

kurosu-being replied:
ご指摘ありがとうございます！
壊れにくいプログラムというお話を改めて宮原さんに解説して頂き、商品を作る為のコーディング
では必ず必要な考え方だととても勉強になりました。
フィールドやプロパティの定義する時、後から変更するのかしないのかという観点でも考える癖をつけたいと思います。

kurosu-being replied:
ご指摘ありがとうございます！
商品コードは一度決めたら変更しないプロパティと判断し、private set; を削除しました。

壊れにくいプログラムというお話を改めて宮原さんに解説して頂き、商品を作る為のコーディング
では必ず必要な考え方だととても勉強になりました。

s-morita-being reviewed (commented):
研修お疲れ様です。
少し見て気になったところをコメントしました。
ご参考まで。
